psh/rmdir/rmdir.c

```diff
@@ -0,0 +1,103 @@
+/*
+ * Phoenix-RTOS
+ *
+ * rmdir - remove empty directories
+ *
+ * Copyright 2023 Phoenix Systems
+ * Author: Gerard Swiderski
+ *
+ * This file is part of Phoenix-RTOS.
+ *
+ * %LICENSE%
+ */
+
+#include <errno.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <unistd.h>
+
+#include "../psh.h"
+
+
+static void psh_rmdir_info(void)
+{
+	printf("remove empty directories");
+}
+
+
+static void psh_rmdir_usage(void)
+{
+	printf("Usage: rmdir [-p] DIRECTORY...\n");
+}
+
+
+static int removedir(char *name, int parents)
+{
+	char *end = name + strlen(name) - (size_t)1;
+	int suppress = 0;
+	int err = -1;
+
+	while (name[0] != '\0') {
+		while ((end >= name) && (*end == '/')) {
+			*(end--) = '\0';
+		}
+
+		if (name[0] != '\0') {
+			err = rmdir(name);
+			end = strrchr(name, '/');
+			if ((err != 0) || (end == NULL) || (parents == 0)) {
+				break;
+			}
```

Comment:
can search '/' from the end and not from begining ( you have a pointer to the end )
e.g. like this:
```suggestion
			while ((end >= name) && (*end != '/')) {
				end--;
			}
			if ((err != 0) || (end <= name) || (parents == 0)) {
				break;
			}
```

Reply:
Yes, I agree your suggestion is better in many aspects, but the resulting binary size is then larger. The version using the `strrchr` call from stdlib/libphoenix still remains smaller. So I'll stick with `strrchr` for now.
Difference: ia32 (48 bytes), imxrt and stm32l4 (32 bytes), risc-V (24 bytes).

btw. As for checking `rmdir -ptr..`, `rm -def..` Your suggestions are ok, still the binary stays smaller than when using e.g. `strcmp`, so I accepted your suggestions.

Reply:
My proposal is faster.
So what is the preferred optimization, for speed or size? I'm starting to get a little lost in this.
If the preference is for optimization for size, then it might be better to use `-Os` instead of `-O2` for compilation.

Reply:
Compact size, not speed: because `psh` is mostly targeted for devices with limited resources. On bigger targets `busybox` should be used. I slowly (in my free time) try to rewrite `psh` and move commonly used routines to one place (not in any pull-request yet), it's a process. When I showed the size differences in previous response, I already used `-Os` (not on ia32).

How frequently `psh` command `rmdir` will be used `is it worth to optimize it for speed or size?` I believe that `resolve_path()` and other system functions are used more frequently and the optimizations should focus there.

Adding new features will increase `psh` size, with no doubt (no full command set should be available on the target  - e.g. no network so no ping, nslookup, ifconfig. The command list may be tailored to resources available on the target). Don't get me wrong. It's not my goal to copy `busybox`, implement it better or something like that (that would be a waste of time), the command set should implement only the bare minimum (simple debugging eg. xd, mem and file operations, I/O, tty, process mgmt, network, date&time, maybe simple scripts), currently missing are only few (e.g. route, ifconfig - as frontend to LwIP used in phoenix-rtos). The `busybox` is still too big for these purposes.